File: src/models/tournament.js
import { Player } from './player.js';
import { Round } from './round.js';

// Tournament entity model with business rules
export class Tournament {
    constructor(name = 'Badminton Tournament') {
        this.id = this.generateId();
        this.name = name;
        this.players = [];
        this.rounds = [];
        this.currentRound = null;
        this.status = 'setup';
        this.createdAt = new Date();
        this.settings = {
            winnerPoints: 2,
            loserPoints: 1
        };
    }

    static fromJSON(data) {
        const tournament = new Tournament(data.name);
        tournament.id = data.id;
        tournament.players = (data.players || []).map(p => Player.fromJSON(p));
        tournament.rounds = (data.rounds || []).map(r => Round.fromJSON(r));
        tournament.currentRound = data.currentRound;
        tournament.status = data.status || 'setup';
        tournament.createdAt = data.createdAt ? new Date(data.createdAt) : new Date();
        tournament.settings = { ...tournament.settings, ...(data.settings || {}) };
        return tournament;
    }

    toJSON() {
        return {
            id: this.id,
            name: this.name,
            players: this.players.map(p => p.toJSON()),
            rounds: this.rounds.map(r => r.toJSON()),
            currentRound: this.currentRound,
            status: this.status,
            createdAt: this.createdAt.toISOString(),
            settings: this.settings
        };
    }

    generateId() {
        return 'tournament_' + Date.now() + '_' + Math.random().toString(36).substr(2, 9);
    }

    // Tournament state transitions
    start() {
        if (this.status !== 'setup') {
            throw new Error('Tournament can only be started from setup state');
        }
        if (this.players.length < 4) {
            throw new Error('Tournament requires at least 4 players to start');
        }
        this.status = 'active';
    }

    complete() {
        if (this.status !== 'active') {
            throw new Error('Tournament can only be completed from active state');
        }
        this.status = 'completed';
    }

    reset() {
        // Reset tournament scores but keep players
        this.players.forEach(player => player.resetScore());
        this.rounds = [];
        this.currentRound = null;
        this.status = 'setup';
    }

    // Player management
    addPlayer(name) {
        if (this.status === 'active') {
            throw new Error('Cannot add players during active tournament');
        }

        // Check for duplicate names
        const existingPlayer = this.players.find(p => 
            p.name.toLowerCase() === name.toLowerCase()
        );
        if (existingPlayer) {
            throw new Error('Player name must be unique');
        }

        const player = new Player(name);
        this.players.push(player);
        return player;
    }

    removePlayer(playerId) {
        if (this.status === 'active') {
            throw new Error('Cannot remove players during active tournament');
        }

        const index = this.players.findIndex(p => p.id === playerId);
        if (index === -1) {
            return false;
        }

        this.players.splice(index, 1);
        return true;
    }

    updatePlayer(playerId, newName) {
        if (this.status === 'active') {
            throw new Error('Cannot update players during active tournament');
        }

        const player = this.getPlayer(playerId);
        if (!player) {
            throw new Error('Player not found');
        }

        // Check for duplicate names (excluding current player)
        const existingPlayer = this.players.find(p => 
            p.id !== playerId && p.name.toLowerCase() === newName.toLowerCase()
        );
        if (existingPlayer) {
            throw new Error('Player name must be unique');
        }

        player.updateName(newName);
        return player;
    }

    getPlayer(playerId) {
        return this.players.find(p => p.id === playerId);
    }

    getPlayers() {
        return [...this.players].sort((a, b) => b.score - a.score || a.name.localeCompare(b.name));
    }

    getActivePlayers() {
        return this.players.filter(p => p.isActive);
    }

    // Round management
    addRound(round) {
        this.rounds.push(round);
        this.currentRound = round.number;
    }

    getCurrentRound() {
        if (this.currentRound === null) {
            return null;
        }
        return this.rounds.find(r => r.number === this.currentRound);
    }

    getRounds() {
        return [...this.rounds].sort((a, b) => a.number - b.number);
    }

    getLastRound() {
        if (this.rounds.length === 0) {
            return null;
        }
        return this.rounds.reduce((latest, round) => 
            round.number > latest.number ? round : latest
        );
    }

    // Validation methods
    canStartTournament() {
        return this.status === 'setup' && this.players.length >= 4;
    }

    canAddRound() {
        if (this.status !== 'active') {
            return false;
        }

        const currentRound = this.getCurrentRound();
        return !currentRound || currentRound.status === 'completed';
    }

    // Scoring methods
    recordMatchResult(matchId, winningPair) {
        const currentRound = this.getCurrentRound();
        if (!currentRound) {
            throw new Error('No active round to record results');
        }

        const match = currentRound.getMatchById(matchId);
        if (!match) {
            throw new Error('Match not found in current round');
        }

        // Record the match result
        match.recordResult(winningPair);

        // Update player scores
        const winningPairIds = match.getWinningPair();
        const losingPairIds = match.getLosingPair();

        winningPairIds.forEach(playerId => {
            const player = this.getPlayer(playerId);
            if (player) {
                player.addScore(this.settings.winnerPoints);
            }
        });

        losingPairIds.forEach(playerId => {
            const player = this.getPlayer(playerId);
            if (player) {
                player.addScore(this.settings.loserPoints);
            }
        });

        // Update round state
        currentRound.recordMatchResult(matchId, winningPair);
    }

    // Rankings and statistics
    getRankings() {
        return this.getPlayers(); // Already sorted by score descending
    }

    getTournamentStats() {
        return {
            totalPlayers: this.players.length,
            totalRounds: this.rounds.length,
            completedRounds: this.rounds.filter(r => r.status === 'completed').length,
            totalMatches: this.rounds.reduce((total, round) => total + round.matches.length, 0),
            completedMatches: this.rounds.reduce((total, round) => 
                total + round.getCompletedMatches().length, 0
            )
        };
    }

    // Settings management
    updateSettings(newSettings) {
        if (this.status === 'active') {
            throw new Error('Cannot modify settings during active tournament');
        }
        this.settings = { ...this.settings, ...newSettings };
    }

    toString() {
        return `Tournament: ${this.name} (${this.status}, ${this.players.length} players, ${this.rounds.length} rounds)`;
    }
}

File: lib/models/tournament.dart
import 'player.dart';
import 'round.dart';

/// Tournament mode enum
enum TournamentMode {
  singles, // 1v1 matches
  doubles; // 2v2 matches

  String get displayName {
    switch (this) {
      case TournamentMode.singles:
        return 'Singles (1v1)';
      case TournamentMode.doubles:
        return 'Doubles (2v2)';
    }
  }

  int get playersPerTeam {
    switch (this) {
      case TournamentMode.singles:
        return 1;
      case TournamentMode.doubles:
        return 2;
    }
  }
}

/// Tournament status enum
enum TournamentStatus {
  setup, // Adding players, not started
  active, // Tournament in progress
  completed; // All matches finished

  String get displayName {
    switch (this) {
      case TournamentStatus.setup:
        return 'Setup';
      case TournamentStatus.active:
        return 'Active';
      case TournamentStatus.completed:
        return 'Completed';
    }
  }
}

/// Tournament model representing the entire tournament
class Tournament {
  final String id;
  final String name;
  final TournamentMode mode;
  final TournamentStatus status;
  final List<Player> players;
  final List<Round> rounds;
  final int winnerPoints;
  final int loserPoints;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Tournament({
    required this.id,
    required this.name,
    required this.mode,
    this.status = TournamentStatus.setup,
    this.players = const [],
    this.rounds = const [],
    this.winnerPoints = 2,
    this.loserPoints = 1,
    DateTime? createdAt,
    this.updatedAt,
  }) : createdAt = createdAt ?? DateTime.now();

  /// Create a Tournament from JSON
  factory Tournament.fromJson(Map<String, dynamic> json) {
    return Tournament(
      id: json['id'] as String,
      name: json['name'] as String,
      mode: TournamentMode.values.firstWhere(
        (m) => m.name == json['mode'],
        orElse: () => TournamentMode.doubles,
      ),
      status: TournamentStatus.values.firstWhere(
        (s) => s.name == json['status'],
        orElse: () => TournamentStatus.setup,
      ),
      players:
          (json['players'] as List?)
              ?.map((p) => Player.fromJson(p as Map<String, dynamic>))
              .toList() ??
          [],
      rounds:
          (json['rounds'] as List?)
              ?.map((r) => Round.fromJson(r as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  /// Convert Tournament to JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'mode': mode.name,
      'status': status.name,
      'players': players.map((p) => p.toJson()).toList(),
      'rounds': rounds.map((r) => r.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  /// Get minimum number of players required
  int get minPlayers {
    switch (mode) {
      case TournamentMode.singles:
        return 2; // At least 2 players for singles
      case TournamentMode.doubles:
        return 4; // At least 4 players for doubles
    }
  }

  /// Check if tournament can start
  bool get canStart =>
      players.length >= minPlayers && status == TournamentStatus.setup;

  /// Check if tournament is in progress
  bool get isActive => status == TournamentStatus.active;

  /// Check if tournament is completed
  bool get isCompleted => status == TournamentStatus.completed;

  /// Get current round number
  int get currentRound => rounds.length;

  /// Create a copy with modified fields
  Tournament copyWith({
    String? id,
    String? name,
    TournamentMode? mode,
    TournamentStatus? status,
    List<Player>? players,
    List<Round>? rounds,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Tournament(
      id: id ?? this.id,
      name: name ?? this.name,
      mode: mode ?? this.mode,
      status: status ?? this.status,
      players: players ?? this.players,
      rounds: rounds ?? this.rounds,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Tournament && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'Tournament{id: $id, name: $name, mode: ${mode.name}, status: ${status.name}, players: ${players.length}, rounds: ${rounds.length}}';
  }
}
